src/crypto-api/index.ts

```diff
@@ -657,38 +657,75 @@ export enum DecryptionFailureCode {
  * Only supported by Rust crypto.
+ *
+ * Message encryption keys are shared with all devices in the room, except in case of
+ * verified user problems (see {@link errorOnVerifiedUserProblems}).
+ *
+ * Events from all senders are always decrypted (and should be decorated with message shields in case
+ * of authenticity warnings, see {@link EventEncryptionInfo}).
  */
-export enum CryptoMode {
-    /**
-     * Message encryption keys are shared with all devices in the room, except for
-     * blacklisted devices, or unverified devices if
-     * `globalBlacklistUnverifiedDevices` is set.  Events from all senders are
-     * decrypted.
-     */
-    Legacy,
+export class NoIsolation {
+    // Discriminated Union
+    public readonly kind: IsolationModeKind.None;
 
     /**
-     * Events are encrypted as with `Legacy` mode, but encryption will throw an error if a
-     * verified user has an unsigned device, or if a verified user replaces
-     * their identity.  Events are decrypted only if they come from cross-signed
-     * devices, or devices that existed before the Rust crypto SDK started
-     * tracking device trust: other events will result in a decryption failure. (To access the failure
-     * reason, see {@link MatrixEvent.decryptionFailureReason}.)
+     * Optional behavior when sharing keys to remote devices.
+     * If set to true, sharing keys will fail (i.e. message sending will fail) with an error if:
+     *     - The user was previously verified but is not anymore.
+     *     - A verified user has some unverified devices (not cross-signed).
+     * If false, the keys will be distributed as usual.
+     * If set to false the client UX should display warnings to inform the user.
      */
-    Transition,
+    public errorOnVerifiedUserProblems: boolean;
 
-    /**
-     * Message encryption keys are only shared with devices that have been cross-signed by their owner.
-     * Encryption will throw an error if a verified user replaces their identity.  Events are
-     * decrypted only if they come from a cross-signed device other events will result in a decryption
-     * failure. (To access the failure reason, see {@link MatrixEvent.decryptionFailureReason}.)
-     */
-    Invisible,
+    public constructor(errorOnVerifiedUserProblems: boolean) {
+        this.kind = IsolationModeKind.None;
+        this.errorOnVerifiedUserProblems = errorOnVerifiedUserProblems;
+    }
 }
 
+/**
+ * A type of device isolation mode used when encrypting or decrypting messages.
+ * Only supported by Rust crypto.
+ *
+ * Message encryption keys are only shared with devices that have been cross-signed by their owner.
+ * Encryption will throw an error if a verified user replaces their identity.
+ *
+ * Events are decrypted only if they come from a cross-signed device other events will result in a decryption
+ * failure. (To access the failure reason, see {@link MatrixEvent.decryptionFailureReason}.)
+ */
+export class OnlySignedIsolation {
```

Comment:
I find the name very long and strange (NoIsolationIsolationMode).
What about [namespaces](https://www.typescriptlang.org/docs/handbook/namespaces.html)? First time I see that, but would allow to do `new DeviceIsolationMode.OnlySigned()` so it's a clear indication that these are `DeviceIsolationMode` without the need of bigger name, also easy to discover?

```
export namespace DeviceIsolationMode {
    export class None {..}
    export class OnlySigned {..}
}
```

Reply:
Per my monologue in #element-dev:matrix.org, I'm not enthusiastic about namespaces, because they seem to be [deprecated](https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html) (and we have a lint rule that enforces that we don't use them).

Reply:
Honestly, I don't find NoIsolationIsolationMode *that bad*, but I agree it's hardly elegant

Reply:
`OnlySignedDevicesIsolationMode` and `AllDevicesIsolationMode` ?

Reply:
Renamed as per andy suggestion, also renamed the `DeviceIsolationModeKind` enum to still match classes names https://github.com/matrix-org/matrix-js-sdk/pull/4429/commits/20f5882b00f05cbef2fd51e7e1bd18c985262447